### PR TITLE
Fix printing ambiguous method group types on invalid casts

### LINKED ISSUE #8556: NullReferenceException while providing diagnostics about ambiguous user defined operators
NRE is thrown while trying assert with the executable below:
SPOT.Transformer.exe -test -type tostring -input .\cs0457-4.cs
cs0457-4.cs:

```
using System;

public delegate void Foo ();

class D
{
	public static implicit operator D (Action d)
	{
		return new D ();
	}

	public static explicit operator D (Foo d)
	{
		return new D ();
	}
}

class Program
{
	static void Main()
	{
		D d = (D) Main;
	}
}

```

Stack trace

```
Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.DiagnosticInfo.AssertMessageSerializable(object[] args) Line 100	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.DiagnosticInfo.DiagnosticInfo(Microsoft.CodeAnalysis.CommonMessageProvider messageProvider, int errorCode, object[] arguments) Line 49	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.DiagnosticInfoWithSymbols.DiagnosticInfoWithSymbols(Microsoft.CodeAnalysis.CSharp.ErrorCode errorCode, object[] arguments, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.CSharp.Symbol> symbols) Line 13	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.CSDiagnosticInfo.CSDiagnosticInfo(Microsoft.CodeAnalysis.CSharp.ErrorCode code, object[] args, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.CSharp.Symbol> symbols, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Location> additionalLocations) Line 32	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.CSDiagnosticInfo.CSDiagnosticInfo(Microsoft.CodeAnalysis.CSharp.ErrorCode code, object[] args) Line 22	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.DiagnosticBagExtensions.Add(Microsoft.CodeAnalysis.DiagnosticBag diagnostics, Microsoft.CodeAnalysis.CSharp.ErrorCode code, Microsoft.CodeAnalysis.Location location, object[] args) Line 35	C#
>	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindCastCore(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax node, Microsoft.CodeAnalysis.CSharp.BoundExpression operand, Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol targetType, bool wasCompilerGenerated, Microsoft.CodeAnalysis.DiagnosticBag diagnostics) Line 1566	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindCast(Microsoft.CodeAnalysis.CSharp.Syntax.CastExpressionSyntax node, Microsoft.CodeAnalysis.DiagnosticBag diagnostics) Line 1528	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindExpressionInternal(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax node, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, bool invoked, bool indexed) Line 428	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindExpression(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax node, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, bool invoked, bool indexed) Line 353	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindValue(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax node, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, Microsoft.CodeAnalysis.CSharp.Binder.BindValueKind valueKind) Line 288	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindPossibleArrayInitializer(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax node, Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol destinationType, Microsoft.CodeAnalysis.DiagnosticBag diagnostics) Line 1997	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindVariableDeclaration(Microsoft.CodeAnalysis.CSharp.Symbols.SourceLocalSymbol localSymbol, Microsoft.CodeAnalysis.CSharp.Symbols.LocalDeclarationKind kind, bool isVar, Microsoft.CodeAnalysis.CSharp.Syntax.VariableDeclaratorSyntax declarator, Microsoft.CodeAnalysis.CSharp.Syntax.TypeSyntax typeSyntax, Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol declTypeOpt, Microsoft.CodeAnalysis.CSharp.Symbols.AliasSymbol aliasOpt, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode associatedSyntaxNode) Line 714	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindVariableDeclaration(Microsoft.CodeAnalysis.CSharp.Symbols.LocalDeclarationKind kind, bool isVar, Microsoft.CodeAnalysis.CSharp.Syntax.VariableDeclaratorSyntax declarator, Microsoft.CodeAnalysis.CSharp.Syntax.TypeSyntax typeSyntax, Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol declTypeOpt, Microsoft.CodeAnalysis.CSharp.Symbols.AliasSymbol aliasOpt, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode associatedSyntaxNode) Line 633	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindDeclarationStatement(Microsoft.CodeAnalysis.CSharp.Syntax.LocalDeclarationStatementSyntax node, Microsoft.CodeAnalysis.DiagnosticBag diagnostics) Line 486	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindStatement(Microsoft.CodeAnalysis.CSharp.Syntax.StatementSyntax node, Microsoft.CodeAnalysis.DiagnosticBag diagnostics) Line 46	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindBlock(Microsoft.CodeAnalysis.CSharp.Syntax.BlockSyntax node, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, Microsoft.CodeAnalysis.CSharp.Binder blockBinder) Line 2050	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindBlock(Microsoft.CodeAnalysis.CSharp.Syntax.BlockSyntax node, Microsoft.CodeAnalysis.DiagnosticBag diagnostics) Line 2036	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.MethodCompiler.BindMethodBody(Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol method, Microsoft.CodeAnalysis.CSharp.TypeCompilationState compilationState, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, out Microsoft.CodeAnalysis.CSharp.ImportChain importChain, out bool originalBodyNested) Line 1497	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethod(Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol methodSymbol, int methodOrdinal, ref Microsoft.CodeAnalysis.CSharp.Binder.ProcessedFieldInitializers processedInitializers, Microsoft.CodeAnalysis.CSharp.SynthesizedSubmissionFields previousSubmissionFields, Microsoft.CodeAnalysis.CSharp.TypeCompilationState compilationState) Line 866	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileNamedType(Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol containingType) Line 432	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileNamedTypeAsTask.AnonymousMethod__0() Line 322	C#
 	Microsoft.CodeAnalysis.dll!Roslyn.Utilities.UICultureUtilities.WithCurrentUICulture.AnonymousMethod__0() Line 136	C#

```

[bin.zip](https://github.com/dotnet/roslyn/files/125250/bin.zip)

### PR DESCRIPTION
**Pull request by OmarTawfik**
Fixes #8556
@dotnet/roslyn-compiler

### CODE REVIEW

**Comment by OmarTawfik** [comment]
@dotnet-bot test this please

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -1784,7 +1784,18 @@ private BoundExpression BindCastCore(ExpressionSyntax node, BoundExpression oper
...
+                    object operandType;
+
+                    if (operand.Kind == BoundKind.MethodGroup)
+                    {
+                        operandType = MessageID.IDS_SK_METHOD.Localize(); // Method groups have null types
+                    }
+                    else
+                    {
+                        operandType = operand.Type;
+                    }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Consider using `?:`: 
`var operandType = (operand.Kind == BoundKind.MethodGroup) ? (object)MessageID.IDS_SK_METHOD.Localize() : operand.Type;`

**Comment by cston** [comment]
LGTM

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-infrastructure is this failure expected? it has been consistent now across a few builds. Tests don't seem to run for some reason (doesn't happen locally).

**Comment by jaredpar** [comment]
@TyOverby 

It looks like another case where the complex CMD escaping in the perf portion of cibuild.cmd is failing builds

> 16:15:25 SUCCESS: The process "VBCSCompiler.exe" with PID 5172 has been terminated.
16:15:25 " was unexpected at this time.

Why is this happening and why is it so inconsistent?

**Comment by TyOverby** [comment]
@jaredpar I have no idea what's happening, and the reason it's so inconsistent is probably due to the special characters in the PR title "(" for example/

**Comment by jaredpar** [comment]
@TyOverby 

> I have no idea what's happening, and the reason it's so inconsistent is probably due to the special characters in the PR title "(" for example/

That both makes sense and is terrifying

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 2]
```diff
@@ -1784,7 +1784,10 @@ private BoundExpression BindCastCore(ExpressionSyntax node, BoundExpression oper
                 ImmutableArray<MethodSymbol> originalUserDefinedConversions = conversion.OriginalUserDefinedConversions;
                 if (originalUserDefinedConversions.Length > 1)
                 {
-                    diagnostics.Add(ErrorCode.ERR_AmbigUDConv, syntax.Location, originalUserDefinedConversions[0], originalUserDefinedConversions[1], operand.Type, targetType);
+                    // Method groups always have null types
+                    var operandType = (operand.Kind == BoundKind.MethodGroup) ? (object)MessageID.IDS_SK_METHOD.Localize() : operand.Type;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: >var operandType = (operand.Kind == BoundKind.MethodGroup) ? (object)MessageID.IDS_SK_METHOD.Localize() : operand.Type; [](start = 20, length = 118)

I think we should simply use operand.Display if operand.Type is null. I think this is the pattern we are following in the codebase, see GetAnonymousTypeFieldType function, for example.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: If the type is not null, I think we should continue using type.

**Pull request merged**